tool/ci/pkg/environment/environment.go

```diff
@@ -0,0 +1,289 @@
+	}
+	if c.Token == "" {
+		return trace.BadParameter("missing parameter token")
+	}
+	if c.unmarshalRevs == nil {
+		c.unmarshalRevs = unmarshalReviewers
+	}
+	return nil
+}
+
+// New creates a new instance of Environment.
+func New(c Config) (*Environment, error) {
+
+	err := c.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	revs, err := c.unmarshalRevs(c.Reviewers, c.Client)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	pr, err := GetPullRequest(c.EventPath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &Environment{
+		Client:           c.Client,
+		reviewers:        revs,
+		defaultReviewers: revs[""],
+		PullRequest:      pr,
+	}, nil
+}
+
+// unmarshalReviewers converts the passed in string representing json object into a map
+func unmarshalReviewers(str string, client *github.Client) (map[string][]string, error) {
+	var hasDefaultReviewers bool
+	if str == "" {
+		return nil, trace.NotFound("reviewers not found")
+	}
+	m := make(map[string][]string)
+
+	err := json.Unmarshal([]byte(str), &m)
+	if err != nil {
+		return nil, err
+	}
+	for author, requiredReviewers := range m {
+		for _, reviewer := range requiredReviewers {
+			_, err := userExists(reviewer, client)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		}
+		if author == "" {
+			hasDefaultReviewers = true
+			continue
+		}
+		_, err := userExists(author, client)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	if !hasDefaultReviewers {
+		return nil, trace.BadParameter("default reviewers are not set. set default reviewers with an empty string as a key")
+	}
+	return m, nil
+
+}
+
+// userExists checks if a user exists
+func userExists(userLogin string, client *github.Client) (*github.User, error) {
+	var targetUser *github.User
+	users, resp, err := client.Search.Users(context.TODO(), userLogin, &github.SearchOptions{})
+	if err != nil || resp.StatusCode != http.StatusOK {
+		return nil, trace.Wrap(err)
+	}
+	for _, user := range users.Users {
+		if *user.Login == userLogin {
+			targetUser = user
+		}
+	}
+	if targetUser == nil {
+		return nil, trace.NotFound("user %s does not exist", userLogin)
+	}
+	return targetUser, nil
+}
+
+// GetReviewersForAuthor gets the required reviewers for the current user.
+func (e *Environment) GetReviewersForAuthor(user string) []string {
+	value, ok := e.reviewers[user]
+	// author is external or does not have set reviewers
+	if !ok {
+		return e.defaultReviewers
+	}
+	return value
+}
+
+// IsInternal determines if an author is an internal contributor.
+func (e *Environment) IsInternal(author string) bool {
+	_, ok := e.reviewers[author]
+	return ok
+}
+
+// GetToken gets token
+func (e *Environment) GetToken() string {
+	return e.token
+}
+
+// GetPullRequest gets the pull request metadata in the current context.
+func GetPullRequest(path string) (*PullRequestMetadata, error) {
+	var actionType action
+	file, err := os.Open(path)
```

Comment:
Doesn't the file contain a lot more than Pull Request data? It feels like you should read that in then extract the Pull Request metadata from that. Something like the following:

```
event, err := ReadEvent(...)
if err != nil {
  return trace.Wrap(err)
}
pr := event.PullRequestMetadata()
if err != nil {
  return trace.Wrap(err)
}
```

Reply:
I am pulling the pull request metadata from the file (event payload). Would you like additional fields pulled? There aren't anymore fields required for the bot to run.